File: src/mcp/services/graph.py
"""
MCP Graph Service

Graph service FastMCP module with 6 tools for knowledge graph operations.
"""

import logging
from mcp.server.fastmcp import FastMCP

logger = logging.getLogger("niko-gateway")

graph_mcp = FastMCP("NikoGraph", stateless_http=True)


def _get_engine():
    """Get graph engine (lazy import to allow test monkeypatching via gateway)."""
    from src.mcp.gateway import get_graph_engine
    return get_graph_engine()


@graph_mcp.tool()
async def graph_query(cypher: str) -> list:
    """
    执行 Cypher 查询

    Args:
        cypher: Cypher 查询语句

    Returns:
        查询结果列表
    """
    engine = _get_engine()
    return await engine.execute_cypher(cypher)


@graph_mcp.tool()
async def graph_get_character(
    name: str,
    include_relations: bool = True,
    include_timeline: bool = False
) -> dict:
    """
    获取角色信息

    Args:
        name: 角色名称
        include_relations: 是否包含关系
        include_timeline: 是否包含时间线

    Returns:
        角色完整信息
    """
    engine = _get_engine()
    return await engine.get_character(name, include_relations, include_timeline)


@graph_mcp.tool()
async def graph_get_relationships(
    character: str,
    relationship_type: str = None,
    depth: int = 1
) -> list:
    """
    获取角色关系网络

    Args:
        character: 角色名称
        relationship_type: 关系类型过滤
        depth: 查询深度

    Returns:
        关系列表
    """
    engine = _get_engine()
    return await engine.get_relationships(character, relationship_type, depth)


@graph_mcp.tool()
async def graph_get_foreshadows(
    status: str = "pending",
    chapter: int = None
) -> list:
    """
    获取伏笔状态

    Args:
        status: 状态过滤 (pending/resolved/abandoned)
        chapter: 章节过滤

    Returns:
        伏笔列表
    """
    engine = _get_engine()
    return await engine.get_foreshadows(status, chapter)


@graph_mcp.tool()
async def graph_add_entity(
    entity_type: str,
    name: str,
    properties: dict = None
) -> dict:
    """
    添加实体到图谱

    Args:
        entity_type: 实体类型 (Character/Location/Event/Item/Foreshadow)
        name: 实体名称
        properties: 属性字典

    Returns:
        {"id": "xxx", "status": "created"}
    """
    engine = _get_engine()
    return await engine.create_entity(entity_type, name, properties or {})


@graph_mcp.tool()
async def graph_add_relation(
    from_name: str,
    to_name: str,
    relation_type: str,
    properties: dict = None
) -> dict:
    """
    添加关系

    Args:
        from_name: 起始实体名称
        to_name: 目标实体名称
        relation_type: 关系类型
        properties: 关系属性

    Returns:
        {"status": "created"}
    """
    engine = _get_engine()
    return await engine.create_relation(from_name, to_name, relation_type, properties or {})


__all__ = ["graph_mcp"]

File: src-ts/mcp/services/graph.ts
/**
 * MCP Graph Service
 *
 * Graph service module with 6 tools for knowledge graph operations.
 * Ported from src/mcp/services/graph.py
 */

import { getGraphEngine } from '../engine';
import { validateEntityType } from '../../utils/cypher-safety.js';

// ---------------------------------------------------------------
// Engine accessor
// ---------------------------------------------------------------

export interface GraphReadScope {
  workspaceId?: string | null;
  projectId?: string | null;
  allowLegacy?: boolean;
}

interface GraphEngine {
  executeCypher(cypher: string, scope?: GraphReadScope | null): Promise<unknown[]>;
  getCharacter(
    name: string,
    includeRelations: boolean,
    includeTimeline: boolean,
    scope?: GraphReadScope | null
  ): Promise<Record<string, unknown>>;
  getRelationships(
    character: string,
    relationshipType?: string | null,
    depth?: number,
    scope?: GraphReadScope | null
  ): Promise<unknown[]>;
  getForeshadows(
    status?: string,
    chapter?: number | null,
    scope?: GraphReadScope | null
  ): Promise<unknown[]>;
  createEntity(
    entityType: string,
    name: string,
    properties: Record<string, unknown>
  ): Promise<Record<string, unknown>>;
  createRelation(
    fromName: string,
    toName: string,
    relationType: string,
    properties: Record<string, unknown>
  ): Promise<Record<string, unknown>>;
}

function isGraphEngine(engine: unknown): engine is GraphEngine {
  return (
    typeof engine === 'object' &&
    engine !== null &&
    typeof (engine as GraphEngine).executeCypher === 'function' &&
    typeof (engine as GraphEngine).getCharacter === 'function' &&
    typeof (engine as GraphEngine).getRelationships === 'function' &&
    typeof (engine as GraphEngine).getForeshadows === 'function' &&
    typeof (engine as GraphEngine).createEntity === 'function' &&
    typeof (engine as GraphEngine).createRelation === 'function'
  );
}

function getEngine(): GraphEngine | null {
  const engine = getGraphEngine();
  return isGraphEngine(engine) ? engine : null;
}

// ---------------------------------------------------------------
// Tool implementations
// ---------------------------------------------------------------

export async function graphQuery(
  cypher: string,
  scope?: GraphReadScope | null
): Promise<unknown[]> {
  const engine = getEngine();
  if (!engine) return [];
  return scope ? engine.executeCypher(cypher, scope) : engine.executeCypher(cypher);
}

export async function graphGetCharacter(
  name: string,
  includeRelations = true,
  includeTimeline = false,
  scope?: GraphReadScope | null
): Promise<Record<string, unknown>> {
  const engine = getEngine();
  if (!engine) return {};
  return scope
    ? engine.getCharacter(name, includeRelations, includeTimeline, scope)
    : engine.getCharacter(name, includeRelations, includeTimeline);
}

export async function graphGetRelationships(
  character: string,
  relationshipType?: string | null,
  depth = 1,
  scope?: GraphReadScope | null
): Promise<unknown[]> {
  const engine = getEngine();
  if (!engine) return [];
  return scope
    ? engine.getRelationships(character, relationshipType, depth, scope)
    : engine.getRelationships(character, relationshipType, depth);
}

export async function graphGetForeshadows(
  state?: string | null,
  chapter?: number | null,
  scope?: GraphReadScope | null
): Promise<unknown[]> {
  const engine = getEngine();
  if (!engine) return [];
  return scope ? engine.getForeshadows(state ?? undefined, chapter, scope) : engine.getForeshadows(state ?? undefined, chapter);
}

export async function graphAddEntity(
  entityType: string,
  name: string,
  properties?: Record<string, unknown> | null
): Promise<Record<string, unknown>> {
  validateEntityType(entityType);
  const engine = getEngine();
  if (!engine) return { error: 'Graph engine unavailable' };
  return engine.createEntity(entityType, name, properties ?? {});
}

export async function graphAddRelation(
  fromName: string,
  toName: string,
  relationType: string,
  properties?: Record<string, unknown> | null
): Promise<Record<string, unknown>> {
  const engine = getEngine();
  if (!engine) return { error: 'Graph engine unavailable' };
  return engine.createRelation(fromName, toName, relationType, properties ?? {});
}
